### PR TITLE
parser: support new COPY syntax

### PR DESCRIPTION
**Pull request by rafiss**
Epic CRDB-18320

Previously, we only supported the COPY syntax used in Postgres versions before 9.0. Since then, Postgres has encouraged this new syntax. Tools such as npgsql and asyncpg use this.

Release note (sql change): Added support for a new syntax to provide options to the COPY statement. The options can now be given in a comma separated list enclosed by parentheses. The old syntax is still supported.

### CODE REVIEW

File: pkg/sql/parser/sql.y [hunk 1]
```diff
@@ -3997,18 +4002,92 @@ copy_options_list:
...
+  }
+| ESCAPE SCONST
+  {
+    $$.val = &tree.CopyOptions{Escape: tree.NewStrVal($2)}
+  }
+| FORCE_QUOTE error
+  {
+    return unimplementedWithIssueDetail(sqllex, 41608, "force_quote")
+  }
+| FORCE_NOT_NULL error
```
Comment (otan) [inline pkg/sql/parser/sql.y:4062]: i don't think this is right:

```
otan=# COPY t FROM STDIN FORCE_NOT_NULL;
ERROR:  syntax error at or near "FORCE_NOT_NULL"
LINE 1: COPY t FROM STDIN FORCE_NOT_NUL
```
vs
```
otan=# COPY t FROM STDIN FORCE NOT NULL id csv;
ERROR:  column "id" of relation "t" does not exist
```

Comment (otan) [reply pkg/sql/parser/sql.y:4062]: hmm, well, the new format supports this:

```
otan=# COPY t FROM STDIN WITH (FORCE_NOT_NULL (id));
ERROR:  COPY force not null available only in CSV mode
```

File: pkg/sql/sem/tree/copy.go [hunk 2]
```diff
@@ -90,17 +95,20 @@ func (o *CopyOptions) Format(ctx *FmtCtx) {
 	var addSep bool
 	maybeAddSep := func() {
 		if addSep {
-			ctx.WriteString(" ")
+			ctx.WriteString(", ")
 		}
 		addSep = true
 	}
+	ctx.WriteString("(")
```
Comment (otan) [inline pkg/sql/sem/tree/copy.go:102]: do you think we should add the `WITH ` for completeness, or nah?

File: pkg/sql/parser/sql.y (hunk 1 above)
Comment (rafiss) [reply pkg/sql/parser/sql.y:4062]: ah nice catch. here is the old grammar (from the bottom of the page at https://www.postgresql.org/docs/current/sql-copy.html)

```
COPY table_name [ ( column_name [, ...] ) ]
    FROM { 'filename' | STDIN }
    [ [ WITH ]
          [ BINARY ]
          [ DELIMITER [ AS ] 'delimiter_character' ]
          [ NULL [ AS ] 'null_string' ]
          [ CSV [ HEADER ]
                [ QUOTE [ AS ] 'quote_character' ]
                [ ESCAPE [ AS ] 'escape_character' ]
                [ FORCE NOT NULL column_name [, ...] ] ] ]

COPY { table_name [ ( column_name [, ...] ) ] | ( query ) }
    TO { 'filename' | STDOUT }
    [ [ WITH ]
          [ BINARY ]
          [ DELIMITER [ AS ] 'delimiter_character' ]
          [ NULL [ AS ] 'null_string' ]
          [ CSV [ HEADER ]
                [ QUOTE [ AS ] 'quote_character' ]
                [ ESCAPE [ AS ] 'escape_character' ]
                [ FORCE QUOTE { column_name [, ...] | * } ] ] ]
```

File: pkg/sql/sem/tree/copy.go (hunk 2 above)
Comment (rafiss) [reply pkg/sql/sem/tree/copy.go:102]: yeah, seems more clear

Comment (rafiss) [reply pkg/sql/sem/tree/copy.go:102]: ah wait, it's already there https://github.com/cockroachdb/cockroach/blob/20e2adda3c76c7172dd986c871df0ae9a346918f/pkg/sql/sem/tree/copy.go#L51

Comment (otan) [reply pkg/sql/sem/tree/copy.go:102]: d'oh

**Review by otan — APPROVED** [review]

**Comment by rafiss** [comment]
tftr!

bors r=otan

**Pull request merged**